File: client/src/main/java/mlyny/model/Sender.java
package mlyny.model;

import java.io.IOException;
import java.util.concurrent.ConcurrentLinkedQueue;

public class Sender extends Thread {
    private final ConcurrentLinkedQueue<Message> m_queue = new ConcurrentLinkedQueue<>();
    private final Client m_client;
    
    public Sender(Client client) {
        m_client = client;
        start();
    }

    public void pushMessage(Message msg) {
        m_queue.add(msg);
    }

    public void run() {
        while(true) {
            while(!m_queue.isEmpty() && m_client.running()) {
                System.out.println("sending message");
                try {
                    m_client.sendMessage(m_queue.remove());
                } catch (IOException e) {
                    System.out.println("failed to send a messsage");
                }
            }

            try {
                Thread.sleep(20);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: server/src/server/sender.cpp
#include "sender.hpp"
#include <sys/socket.h>

void Sender::push_message(Message msg) {
    m_queue.push(std::move(msg));
}

void Sender::run(Sender* sender) {
    while(true) {
        while(!sender->m_queue.empty()) {
            Message msg = sender->m_queue.pop();
            // TODO: send the msg
            send(msg.socket(), msg.data().data(), msg.data().size(), 0);
        }
    }
}
